filebeat/channel/util.go

```diff
@@ -60,7 +63,13 @@ func SubOutlet(out Outleter) Outleter {
 func (o *subOutlet) Close() error {
 	isOpen := o.isOpen.Swap(false)
 	if isOpen {
+		// Signal OnEvent() to terminate
 		close(o.done)
+		// This mutex prevents the event channel to be closed if OnEvent is
+		// still running.
+		o.mutex.Lock()
+		defer o.mutex.Unlock()
+		close(o.ch)
```

Comment:
as @andrewkroh noted `sync.Once` could replace the semaphore **isOpen**.

Reply:
The open flag is also used in `OnEvent`. Once the outlet is closed, we MUST NOT run into the select block in OnEvent. A `select` is non-deterministic. Even if o.done is closed, we might end up sending something to `o.ch`. Depending on current state this can generate either a panic or some inconsistent state .